### PR TITLE
Add Stroke Width Transform algorithm for Text Detection

### PR DESCRIPTION
**Pull request by devanshbatra04**
closes https://github.com/opencv/opencv/issues/5506

Please don't merge yet. I will be adding test and sample code. @berak can you kindly have a look? I feel I am including everything but I get an error. OpenCV compiles perfectly (or so I suppose).

```undefined reference to `cv::text::SWTTextDetection(cv::Mat const&, bool)'``` when I compile a sample program. The hpp file is loaded but apparently cmake or gcc cannot find SWTTextDetection.cpp. I have been stuck on this for two days now :/
I could confirm that SWTTextDetection.cpp is definitely being compiled by cmake.

This most definitely has something to do with SWTTextDetection.cpp not being loaded or linked during make. 

<cut/>
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,windows,mac,docs
```

### CODE REVIEW

File: modules/text/include/opencv2/text/swtTextDetection.hpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#ifndef SWT_H
+#define SWT_H
+
+#include <opencv2/core.hpp>
+#include <vector>
+
+namespace cv {
+namespace text {
+/** @brief Applies the Stroke Width Transform operator followed by filtering of connected components of similar Stroke Widths to find letter candidates and chain them by proximity and size.
+    @param input_image the input image with 3 channels.
```
Comment (berak) [inline modules/text/include/opencv2/text/swtTextDetection.hpp]: the actual algorithm works on single channel grayscale  (and the multi-channel image is used , to draw visualizations into it) right ?

while being very useful, imho those visualizations  should not be mandatory.

what about this idea: pass an optional drawing image as another argument, like:

    void SWTTextDetection (InputArray input_image, bool dark_on_light, OutputArray draw=noArray()) {
        if (draw.needed()) {
                // if they passed in the color image, use it,
                // if they passed in an empty Mat, create a 3chan grayscale image and draw into that
        }
    }

**Comment by berak** [comment]
[doc errors](https://pullrequest.opencv.org/buildbot/builders/precommit-contrib_docs/builds/13411/steps/whitespace%20opencv_contrib/logs/stdio) : if you rename `opencv_contrib/.git/hooks/pre-commit.sample` to `opencv_contrib/.git/hooks/pre-commit` , you will get the warnings locally at  commit time (and can fix them before pushing)

File: modules/text/include/opencv2/text/swtTextDetection.hpp (hunk 1 above)
Comment (devanshbatra04) [reply modules/text/include/opencv2/text/swtTextDetection.hpp]: > the actual algorithm works on single channel grayscale

Most of it yes, there is a part where the color averages are taken into account. The paper assumes two consecutive letter candidates to have a similar color average. So color image is pretty much mandatory.

> pass an optional drawing image as another argument

Awesome, I will do that. thanks!

Comment (berak) [reply modules/text/include/opencv2/text/swtTextDetection.hpp]: > (the findValidChains function)

ah, right. missed it.

**Comment by devanshbatra04** [comment]
Is there any documentation on writing tests? It doesn't seem as trivial.

**Comment by savuor** [comment]
> Is there any documentation on writing tests? It doesn't seem as trivial.

Sure, check [this](https://github.com/opencv/opencv/wiki/Coding_Style_Guide) and [this](https://github.com/opencv/opencv/wiki/QA_in_OpenCV) for example

**Comment by devanshbatra04** [comment]
Here's the output for the sample program. Please let me know what else is required.
Screenshot from 2020-03-29 05-59-14

File: modules/text/samples/textdetection_swt.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        cout << "Insuficient parameters. Aborting!" << endl;
+        cout << "Usage: textdetection_swt file_path dark_on_light" << endl;
+        cout << "Usage: To detect dark text on light background pass dark_on_light as 1, otherwise pass 0" << endl;
+        cout << "Example: ./swt ./scenetext_segmented_word02.jpg 0" << endl;
+        exit(1);
+    }
+
+    bool dark_on_light = (argv[2][0] != '0');
+
+    Mat image = imread("/home/opencv-dev/OSS/opencv_contrib/modules/text/samples/scenetext_segmented_word03.jpg", IMREAD_COLOR);
```
Comment (alalek) [inline modules/text/samples/textdetection_swt.cpp]: Absolute paths are not allowed. Use command line arguments to specify input file.

Consider using OpenCV's CommandLineParser to handle that (like `segmented_word_recognition.cpp`).

File: modules/text/samples/textdetection_swt.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    namedWindow( "Input Image", WINDOW_AUTOSIZE );
+    imshow("Input Image", image);
```
Comment (alalek) [inline modules/text/samples/textdetection_swt.cpp]: `namedWindow()` call is not need. `imshow()` is enough.

File: modules/text/samples/textdetection_swt.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    cout << "Starting SWT Text Detection Demo with Dark on Light set to 0" << endl;
```
Comment (alalek) [inline modules/text/samples/textdetection_swt.cpp]: `dark_on_light` variable?

File: modules/text/samples/textdetection_swt.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    vector<cv::Rect> components;
+    Mat out( image.size(), CV_8UC3 );
```
Comment (alalek) [inline modules/text/samples/textdetection_swt.cpp]: Destination matrices in OpenCV should be allocated to processing code (`detectTextSWT` in this case)

File: modules/text/src/text_detector_swt.cpp [hunk 6]
```diff
@@ -0,0 +1,812 @@
...
+    for (unsigned int i = 0; i < adj[v].size(); i++) {
+        int neighbour = adj[v][i];
+        if (!visited[neighbour]) {
+            DFSUtil(neighbour, visited, adj, label, component_id);
+        }
+    }
+}
+
+int connected_components(std::vector<int> adj[], std::vector<int> &component_id, int num_vertices) {
+    bool *visited = new bool[num_vertices];
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp]: Manual memory managements must be avoided - new/delete, malloc/free.

Use C++ `std::vector` or OpenCV's `cv::AutoBuffer` instead.

File: modules/text/src/text_detector_swt.cpp [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+
+
+    // Convert to grayscale
+    Mat grayImage(input.size(), CV_8UC1);
+    cvtColor(input, grayImage, COLOR_RGB2GRAY);
+    // Create Canny Image
+    double threshold_low = 175;
+    double threshold_high = 320;
+    Mat canny_edge_image(input.size(), CV_8UC1);
+    Canny (grayImage, canny_edge_image, threshold_low, threshold_high, 3);
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp]: BTW, There is `Canny()` overloads with custom pre-computed dx/dy gradients.

> Mat canny_edge_image(input.size(), CV_8UC1);

Avoid allocation of result matrices.

File: modules/text/src/text_detector_swt.cpp [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+}
+}
+
+CV_EXPORTS_W void detectTextSWT (InputArray input, CV_OUT std::vector<cv::Rect>& result, bool dark_on_light, OutputArray & draw /*=noArray()*/, OutputArray & chainBBs /*=noArray()*/) {
+    CV_CheckTypeEQ(input.type(), CV_8UC3, "");
+
+
+    // Convert to grayscale
+    Mat grayImage(input.size(), CV_8UC1);
+    cvtColor(input, grayImage, COLOR_RGB2GRAY);
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp]: > COLOR_RGB2GRAY

Default OpenCV color space is BGR (not RGB).

File: modules/text/test/test_detection_swt.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+#include "test_precomp.hpp"
+#include "opencv2/imgcodecs.hpp"
+
+using namespace cv;
+using namespace std;
+namespace opencv_test { namespace {
+
+TEST (TextDetectionSWT, accuracy_light_on_dark) {
+    std::vector<cv::Rect> letters;
+    String dataPath = cvtest::TS::ptr()->get_data_path() + "cv/mser/mser_test.png";
```
Comment (alalek) [inline modules/text/test/test_detection_swt.cpp]: Please use `cvtest::findDataFile()` instead (see `test_detection.cpp`)

You may use images from samples directory too.

File: modules/text/test/test_detection_swt.cpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    Mat image = imread(dataPath, IMREAD_COLOR);
+    vector<cv::Rect> components;
+    cv::text::detectTextSWT(image, components, false);
+    cout << components.size();
+    /* all 5 letter candidates should be identified (R9888) */
+    EXPECT_EQ(components.size(), (unsigned) 5);
```
Comment (alalek) [inline modules/text/test/test_detection_swt.cpp]: Swap arguments: `EXPECT_EQ(expected_value, actual_value)`

File: modules/text/test/test_detection_swt.cpp [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+using namespace std;
+namespace opencv_test { namespace {
+
+TEST (TextDetectionSWT, accuracy_light_on_dark) {
+    std::vector<cv::Rect> letters;
+    String dataPath = cvtest::TS::ptr()->get_data_path() + "cv/mser/mser_test.png";
+    Mat image = imread(dataPath, IMREAD_COLOR);
+    vector<cv::Rect> components;
+    cv::text::detectTextSWT(image, components, false);
+    cout << components.size();
```
Comment (alalek) [inline modules/text/test/test_detection_swt.cpp]: No need to dump this value. EXPECT_EQ below will do that.

File: modules/text/test/test_detection_swt.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+#include "test_precomp.hpp"
+#include "opencv2/imgcodecs.hpp"
+
+using namespace cv;
+using namespace std;
+namespace opencv_test { namespace {
+
+TEST (TextDetectionSWT, accuracy_light_on_dark) {
+    std::vector<cv::Rect> letters;
```
Comment (alalek) [inline modules/text/test/test_detection_swt.cpp]: > letters

not used.

And below (all 5 cases).

File: modules/text/test/test_detection_swt.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    String dataPath = cvtest::TS::ptr()->get_data_path() + "cv/mser/mser_test.png";
+    Mat image = imread(dataPath, IMREAD_COLOR);
+    vector<cv::Rect> components;
+    cv::text::detectTextSWT(image, components, false);
```
Comment (alalek) [inline modules/text/test/test_detection_swt.cpp]: > `cv::`

This prefix is usually not need in OpenCV source code (except conflicts)

File: modules/text/src/text_detector_swt.cpp [hunk 14]
```diff
@@ -0,0 +1,812 @@
...
+    // the inner vector contains the (y,x) of each pixel in that component.
+    std::vector<std::vector<SWTPoint> > components = getComponents(SWTImage);
+    std::vector<Component> validComponents = filterComponents(SWTImage, components, false);
+    Mat outDrawing( input.size(), CV_8UC3 );
+    if (draw.needed()) {
+        outDrawing = draw.getMat();
+    }
+    Mat input_image = input.getMat();
+    vector<cv::Rect> tempVec;
+    std::vector<cv::Rect>& outTextRegions = chainBBs.isVector() ? *reinterpret_cast<vector<cv::Rect> *>(chainBBs.getObj()): tempVec;
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp]: Please use this pattern:
```
vector<Rect> outTextRegions;
... compute it ...
if (chainBBs.isNeeded())  // to handle noArray()
{
    _InputArray(outTextRegions).copyTo(chainBBs);  // or Mat(outTextRegions).copyTo(chainBBs);
}
```

File: modules/text/src/text_detector_swt.cpp [hunk 15]
```diff
@@ -0,0 +1,812 @@
...
+        for (unsigned int i = 0; i < chains.size(); i++) {
+            chains[i].merged = false;
+        }
+        merges = 0;
+        std::vector<ChainedComponent> chainsAfterMerging;
+        for (unsigned int i = 0; i < chains.size(); i++) {
+            for (unsigned int j = 0; j < chains.size(); j++){
+                if (i!=j && !chains[i].merged && !chains[j].merged) {
+                    if (chains[i].chainIndexA == chains[j].chainIndexA) {
+                        if (acos(chains[i].dir.x * -chains[j].dir.x + chains[i].dir.y * -chains[j].dir.y) < alignmentThreshold) {
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp]: `if (acos(...expr...) < alignmentThreshold)`

if expr > 0 then:

```
const double alignmentThreshold_cos = cos(alignmentThreshold); // cos(PI/6) = 0.5 btw
...
if (expr > alignmentThreshold_cos)
```

can be used.

File: modules/text/samples/textdetection_swt.cpp [hunk 16]
```diff
@@ -0,0 +1,65 @@
+// // This file is part of OpenCV project.
+// // It is subject to the license terms in the LICENSE file found in the top-level directory
+// // of this distribution and at http://opencv.org/license.html.
+#include <opencv2/text.hpp>
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/dnn.hpp>
```
[the remark this answers is not in the document]
Comment (devanshbatra04) [reply modules/text/samples/textdetection_swt.cpp:7]: imgproc is used, to draw the rectangles. removing dnn.

File: modules/text/samples/textdetection_swt.cpp (hunk 5 above)
Comment (devanshbatra04) [reply modules/text/samples/textdetection_swt.cpp:45]: I am assuming that means declaring Mat out in sample but actually allocating the space in the ```detectTextSWT```?

File: modules/text/test/test_detection_swt.cpp (hunk 9 above)
Comment (devanshbatra04) [reply modules/text/test/test_detection_swt.cpp:14]: Having issues with this. It doesn't work on my build, not even in ```test_detection.cpp```. I will try to debug

File: modules/text/samples/textdetection_swt.cpp [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+        exit(1);
+    }
+
+    string filepath = argv[1];
+    if (!fileExists(filepath)) {
+        cout << "Error: The input image does not exist" << endl;
+        cout << "Please check the path to image file." << endl;
+        cout << "Usage: textdetection_swt file_path dark_on_light" << endl;
+        cout << "Usage: To detect dark text on light background pass dark_on_light as 1, otherwise pass 0" << endl;
+        cout << "Example: ./swt ./scenetext_segmented_word02.jpg 0" << endl;
```
Comment (alalek) [inline modules/text/samples/textdetection_swt.cpp:39]: Please use `cv::CommandLineParser` in OpenCV samples

File: modules/text/src/text_detector_swt.cpp [hunk 18]
```diff
@@ -0,0 +1,817 @@
...
+        }
+        chains = newchains;
+        std::stable_sort(chains.begin(), chains.end(), chainSortLength);
+    }
+
+    std::vector<ChainedComponent> newchains;
+    std::vector<std::vector<SWTPoint>> componentsPointsVector;
+    vector<Component> finalComponents;
+    finalComponents.reserve(components.size());
+    bool componentIncluded [components.size()] = {false};
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp:719]: - `cv::AutoBuffer`
- `std::vector`

File: modules/text/test/test_detection_swt.cpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "test_precomp.hpp"
+#include "opencv2/imgcodecs.hpp"
+
+using namespace cv;
+using namespace std;
```
Comment (alalek) [inline modules/text/test/test_detection_swt.cpp:9]: Not needed at due "namespace opencv_test"

File: modules/text/src/text_detector_swt.cpp [hunk 20]
```diff
@@ -0,0 +1,817 @@
...
+    Mat outTemp( output.size(), CV_32FC1 );
+    renderComponents(SWTImage, finalComponents, outTemp);
+    outTemp.convertTo(outTemp, CV_8UC1, 255.);
+    cvtColor (outTemp, output, COLOR_GRAY2RGB);
+    renderComponentBBs(finalComponents, output);
+    return getComponentBBs(finalComponents);
+}
+}
+
+CV_EXPORTS_W void detectTextSWT (InputArray input, CV_OUT std::vector<cv::Rect>& result, bool dark_on_light, OutputArray & draw /*=noArray()*/, OutputArray & chainBBs /*=noArray()*/) {
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp:760]: > CV_EXPORTS_W

Should be removed from .cpp files (just `CV_EXPORTS` too).

File: modules/text/src/text_detector_swt.cpp [hunk 21]
```diff
@@ -0,0 +1,817 @@
...
+#include "precomp.hpp"
+#include "opencv2/core.hpp"
+#include "opencv2/imgproc.hpp"
+#include <cmath>
+#include <math.h>
+#include <utility>
+#include <iostream>
+#include <algorithm>
+#include <unordered_map>
+#include <vector>
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp:13]: Please filter out headers which are already in precomp.hpp or not used (like `iostream`)

**Comment by devanshbatra04** [comment]
A test is failing in the windows build. 

```unknown file: error: SEH exception with code 0xc00000fd thrown in the test body.```
I don't understand what this means.

**Comment by devanshbatra04** [comment]
@alalek the builds are passing are passing but Github is still waiting for the builds, not sure why. 

I have made all the edits. Can you please see if anything else is required? I really want to get this merged.

**Comment by alalek** [comment]
> 0xc00000fd

Stack overflow exception.

> waiting for the builds

Not all required builds are run (due source branch name - "master" prevents triggering of builds automatically). I scheduled them.

File: modules/text/src/text_detector_swt.cpp [hunk 22]
```diff
@@ -0,0 +1,854 @@
...
+        if (!skipChecks && attributes.variance > 0.5 * attributes.mean) continue;
+        if (!skipChecks && attributes.width > 300) continue;
+
+
+        float area = attributes.length * attributes.width;
+
+        // compute the rotated bounding box
+        for (int theta_i = 1; theta_i < (NUM_THETA / 2); theta_i++)
+        {
+            float theta = (float)i * (CV_PI / NUM_THETA);
```
Comment (alalek) [inline modules/text/src/text_detector_swt.cpp:446]: Replaced for loop with floating-point iterator to integer one (floating-point is not accuracy between platforms and should be avoided).
Please check loop boundary (`<` pr `<=`) and fix if necessary.

Old code:
```
float increment = ((float) 1. )/ (float) 36.;
for (float theta = increment * (float) CV_PI; theta<CV_PI/2.0; theta += increment * (float) CV_PI) {
```

Comment (devanshbatra04) [reply modules/text/src/text_detector_swt.cpp]: The `<` looks good. But in line 446, it should `(float) theta_i * (CV_PI / NUM_THETA)` and not  `(float)i * (CV_PI / NUM_THETA)`. I will correct that and go over the rest of the code.

**Pull request merged**